src/backend/distributed/utils/tuplestore.c

```diff
@@ -0,0 +1,98 @@
+/*-------------------------------------------------------------------------
+ *
+ * tuplestore.c
+ *	  Utilities regarding calls to PG functions
+ *
+ * Copyright (c) 2012-2018, Citus Data, Inc.
+ *-------------------------------------------------------------------------
+ */
+
+
+#include "postgres.h"
+
+#include "distributed/tuplestore.h"
+#include "miscadmin.h"
+
+/*
+ * CheckTuplestoreReturn checks if a tuplestore can be returned in the callsite
+ * of the UDF.
+ */
+ReturnSetInfo *
+CheckTuplestoreReturn(FunctionCallInfo fcinfo, TupleDesc *tupdesc)
+{
+	/* IMPORTANT can this be undefined behaviour. Casting to ReturnSetInfo before
+	 * we checked the type with IsA?
+	 * In case the type actual type is more strictly aligned than ReturnSetInfo
+	 * it is it seems:
+	 * https://wiki.sei.cmu.edu/confluence/display/c/EXP36-C.+Do+not+cast+pointers+into+more+strictly+aligned+pointer+types */
+	ReturnSetInfo *returnSetInfo = (ReturnSetInfo *) fcinfo->resultinfo;
```

Comment:
If this is indeed a problem it's easily fixable by moving it after the check. I mainly want to know if this is indeed a possible problem, because it was done like this everywhere before.

Reply:
Doing the IsA check later seems safe. We're just not allowed to look at any fields that are not in Node until we did the check.

The problem in the examples is that an "int" is 4 bytes while a char is 1 byte. Originally, the pointer was pointing at a 1 byte value, while after the cast it's pointing at 4 byte value and that changes the way the pointer is resolved.

Reply:
The way I read it, was that dereferencing/accessing fields on the pointer doesn't even matter. In the first example the pointer casting alone is already a making the pointers unequal. So I would assume you can still get the same problem in this case if fields further down the struct definitions of the two types don't align the same.

e.g.:
```c
struct A {
   uint64 shared,
   char a,
}
struct B {
   uint64 shared,
   int b,
}

void main() {
    struct A = struct A{
}
```

Reply:
That seems kind of a silly thing for a compiler to do even for big endian architectures. Anyway, pointers to structs point to the first member of the struct, which does not change here. Postgres uses this pattern all over the place.

Reply:
Alright, I'll keep it like this since Postgres also does it like this too. Modern compilers make it a habbit of doing silly things when you hit undefined behaviour though. See this for a horrible example: https://gcc.godbolt.org/z/d9eCOh

One thing I could imagine happening is that it optimizes away the `IsA` check here, because everytime when this specific layout is created it also sets the field to `ReturnSetInfo`.